src/main/java/net/minecraftforge/registries/RegisterEvent.java

```diff
@@ -0,0 +1,89 @@
+package net.minecraftforge.registries;
+
+import net.minecraft.core.Registry;
+import net.minecraft.resources.ResourceKey;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraftforge.eventbus.api.Event;
+import net.minecraftforge.fml.event.IModBusEvent;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * This event fires when all registries are ready to have modded objects registered.
+ * All registries will be ready and unfrozen when this event fires.
+ * <p>
+ * Fired on the {@link IModBusEvent mod bus}.
+ *
+ * @see #register(ResourceKey, ResourceLocation, Object)
+ */
+public class RegisterEvent extends Event implements IModBusEvent
+{
+    @NotNull
+    private final ResourceKey<? extends Registry<?>> registryKey;
+    @Nullable
+    final ForgeRegistry<?> forgeRegistry;
+    @Nullable
+    private final Registry<?> vanillaRegistry;
+
+    RegisterEvent(@NotNull ResourceKey<? extends Registry<?>> registryKey, @Nullable ForgeRegistry<?> forgeRegistry, @Nullable Registry<?> vanillaRegistry)
+    {
+        this.registryKey = registryKey;
+        this.forgeRegistry = forgeRegistry;
+        this.vanillaRegistry = vanillaRegistry;
+    }
+
+    /**
+     * Registers the value with the given name to the stored registry if the provided registry key matches this event's registry key.
+     *
+     * @param registryKey The key of the registry to register the value to
+     * @param name The name of the object to register as its key
+     * @param value The object value
+     * @param <T> The type of the registry
+     */
```

Comment:
The test mods show that the convention for registering using this event is to first check if the registry key in this event matches the key they expect, and calling this event if so. To my understanding, this is to avoid the creation of the to-be-registered object.

This convention should be documented in the javadoc of this class. Additionally or alternatively, a similar method could be added which accepts a `Supplier<T>` (though, this will still incur the cost of the `ResourceLocation` creation).

Reply:
After changing this method to take a supplier and providing a RegisterHelper, I don't think this is important to specify anymore.